### DISCUSSION TITLE
开启YLT_ENABLE_SSL 并用openssl生成文件连接失败

### DISCUSSION THREAD

**Discussion by jjqcat** [alibaba/yalantinglibs#694] [Q&A]
我按下面步骤产生 server.key，server.csr， server.crt，客户端连接失败，不知道什么原因，但用demo里面的文件是可以的。

openssl genrsa -des3 -out server.key 4096

openssl req -new -key server.key -out server.csr

openssl x509 -req -days 3650 -in server.csr -signkey server.key -out server.crt

openssl dhparam -out dh512.pem 512

openssl dhparam -out dhparam.pem 4096

**Comment by qicosmos** [comment]
rpc 还是http？

**Comment by jjqcat** [comment]
RPC

**Comment by qicosmos** [comment]
生成证书时密码设置了吗？

**Comment by jjqcat** [comment]
设置了

**Comment by jjqcat** [comment]
客户端跟踪是握手失败，服务端，客户端初始化都是成功的

**Comment by qicosmos** [comment]
你先把密码改成test试试。
现在rpc内部校验的passwd写死成了test

**Comment by jjqcat** [comment]
好，我试下

**Comment by jjqcat** [comment]
好像还是不行

**Comment by qicosmos** [comment]
把512改成2048试试

**Comment by jjqcat** [comment]
dh512.pem dhparam.pem 两个文件内部有用到吗

**Comment by jjqcat** [comment]
还是连接不上

**Comment by qicosmos** [comment]
你的openssl是什么版本？

**Comment by jjqcat** [comment]
3.2.1

**Comment by qicosmos** [comment]
好的，稍晚点测试查一下。

**Comment by jjqcat** [comment]
好的，谢谢。 我用 1024，产生server.key, 初始化报错EE key too small

**Comment by jjqcat** [comment]
用demo提供的server.key，server.csr,server.crt则没有问题

**Comment by poor-circle** [comment]
会不会是长度问题？试下2048？

> 好的，谢谢。 我用 1024，产生server.key, 初始化报错EE key too small

**Comment by jjqcat** [comment]
这个是长度问题，改成2048不报错了，现在就是连不上的问题

**Comment by jjqcat** [comment]
[openssl_3.2.1.zip](https://github.com/user-attachments/files/15614297/openssl_3.2.1.zip)

**Comment by jjqcat** [comment]
这个工具是我编译的，里面有我测试生成脚本，麻烦抽空验证下，谢谢

**Comment by qicosmos** [comment]
好的，要稍晚点哈。

**Comment by jjqcat** [comment]
麻烦确认下这个问题，谢谢

**Comment by qicosmos** [comment]
这两天看一下这个问题。

**Comment by jjqcat** [comment]
好的，再次感谢

**Comment by qicosmos** [comment]
目前还没有windows环境，我在centos7 机器上安装并验证了openssl3.2.1，过程如下，供你参考：

## 生成server.key
输入：
```
openssl genrsa -des3 -out server.key 4096
```

提示：
`Enter PEM pass phrase:`
这里必须要输入4个以上的字符，输入`test`

接着会继续提示：
```
Verifying - Enter PEM pass phrase:
```
输入`test`

## 生成server.csr
输入：
```cpp
openssl req -new -key server.key -out server.csr
```
会出现下列输入提示：
```cpp
Enter pass phrase for server.key:
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [AU]:
State or Province Name (full name) [Some-State]:
Locality Name (eg, city) []:
Organization Name (eg, company) [Internet Widgits Pty Ltd]:
Organizational Unit Name (eg, section) []:
Common Name (e.g. server FQDN or YOUR name) []:
Email Address []:

Please enter the following 'extra' attributes
to be sent with your certificate request
A challenge password []:
An optional company name []:
```
在`Enter pass phrase for server.key:` 这一步输入`test`

之后一路回车到最后一步。

## 生成server.key

输入：
```cpp
openssl x509 -req -days 3650 -in server.csr -signkey server.key -out server.crt
```
提示输入：
```
Enter pass phrase for server.key:
```
输入`test`

显示：
```
Certificate request self-signature ok
subject=C=AU, ST=Some-State, O=Internet Widgits Pty Ltd
```

## coro_rpc 使用生成openssl文件

coro_rpc_server使用刚生成的server.crt和server.key文件:

```cpp
server.init_ssl(
      ssl_configure{"../openssl_files", "server.crt", "server.key"});
```

coro_rpc_client使用刚生成的server.key文件:
```cpp
client->init_ssl("../openssl_files", "server.crt");
co_await client->connect("127.0.0.1", port);
```

注意在编译时要设置YLT_ENABLE_SSL为ON，以上测试可以正常连接到服务端。

openssl 测试代码在src/coro_rpc/tests/test_coro_rpc_client.cpp文件中。

**Comment by jjqcat** [comment]
好的，谢谢，我马上验证下

**Comment by jjqcat** [comment]
windows下还是报错，相关操作步骤和提示跟上面提到的一样，服务端报错：  [coro_connection.hpp:168] handshake failed: tlsv1 alert internal error (SSL routines) conn_id 1； 客户端报错： [impl/coro_rpc_client.hpp:443] client_id 0 handshake failed: certificate verify failed (SSL routines)

**Comment by qicosmos** [comment]
你在Linux 也测试一下看看。

**Comment by jjqcat** [comment]
能否用windows下生成的几个文件到centos7试下？

**Comment by qicosmos** [comment]
这个要晚点找个环境试一下才行。

为啥一定要在windows上，你的部署环境是windows吗？

**Comment by jjqcat** [comment]
是的，我是在windows系统下用

**Comment by jjqcat** [comment]
有办法在windows系统下测试吗，或者是否可以用我在windows下产生的server.crt，server.key,server.csr等文件，在centos7下确认下？可以的话我上传下，谢谢

**Comment by poor-circle** [comment]
> 有办法在windows系统下测试吗，或者是否可以用我在windows下产生的server.crt，server.key,server.csr等文件，在centos7下确认下？可以的话我上传下，谢谢

我稍后会测试一下

**Comment by jjqcat** [comment]
好的，谢谢

**Comment by jjqcat** [comment]
[openssl_files.zip](https://github.com/user-attachments/files/15832823/openssl_files.zip)
麻烦用这个windows平台下3.2.1产生的openssl files测试下，谢谢

**Comment by jjqcat** [comment]
在windows平台下，找第三方编译的openssl也尝试过，还是报一样的错误，不知道是证书问题还是其它BUG？直接把生成的证书相关文件放在demo里面，也是报同样的错误，还请帮忙验证下，谢谢

**Comment by jjqcat** [comment]
已经解决 openssl x509 命令需要增加参数 -subj "/CN=localhost"

**Comment by qicosmos** [comment]
但是linux下不需要加这个参数啊
